core/src/main/java/org/elasticsearch/common/Numbers.java

```diff
@@ -178,4 +181,54 @@ public static boolean isValidDouble(double value) {
         }
         return true;
     }
+
+    /** Return the long that {@code n} stores, or throws an exception if the
+     *  stored value cannot be converted to a long that stores the exact same
+     *  value. */
+    public static long toLongExact(Number n) {
+        if (n instanceof Byte || n instanceof Short || n instanceof Integer
+                || n instanceof Long || n instanceof Float || n instanceof Double) {
+            double d = n.doubleValue();
```

Comment:
I think it'd be a bit cleaner not to do this for Byte, Short, Integer, and Long.